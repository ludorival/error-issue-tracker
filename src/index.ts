'use strict'

export * from './trackErrors'

export type Error = {
  message: string
  timestamp: number
  [key: string]: unknown
}

export type Issue = {
  id: string
  url: string
  [key: string]: unknown
}
export interface FederatedErrors {
  id?: string
  name: string
  projectId: string
  firstOccurrenceTimeStamp: number
  lastOccurrenceTimeStamp: number
  occurrences: Error[]
}

export type TrackedErrors = FederatedErrors & { issue: Issue }

export type SavedTrackedErrors = TrackedErrors & { id: string }
export interface ErrorDatabase {
  save(error: TrackedErrors): Promise<SavedTrackedErrors>
  fetch(projectId: string): Promise<SavedTrackedErrors[]>
}
export interface IssueClient {
  createIssue(error: FederatedErrors): Promise<Issue>
  updateIssue(error: SavedTrackedErrors): Promise<Issue>
}
export type Comparator = (source: Error, target: Error) => number
export type TrackErrorOptions = {
  database: ErrorDatabase
  issueClient: IssueClient
  projectId: string
  compareError?: Comparator
}
