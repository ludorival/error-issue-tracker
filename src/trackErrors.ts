import {
  Error,
  FederatedErrors,
  SavedTrackedErrors,
  TrackErrorOptions,
} from '.'

export async function trackErrors(
  errors: Error[],
  {
    database,
    issueClient,
    projectId,
    compareError = (source, target) =>
      source.message.localeCompare(target.message),
  }: TrackErrorOptions
): Promise<SavedTrackedErrors[]> {
  const compareGroupedError = (
    source: FederatedErrors,
    target: FederatedErrors
  ) => compareError(source.occurrences[0], target.occurrences[0])
  const groupedErrors = zipErrors(
    errors.map<FederatedErrors>(
      (error) =>
        ({
          name: error.message,
          firstOccurrenceTimeStamp: error.timestamp,
          lastOccurrenceTimeStamp: error.timestamp,
          projectId,
          occurrences: [error],
        } as FederatedErrors)
    ),
    compareGroupedError
  )
  const savedErrors = await database.fetch(projectId)
  const newErrors = zipErrors(
    savedErrors.concat(groupedErrors as SavedTrackedErrors[]),
    compareGroupedError
  )
  const newSavedErrors = await Promise.all(
    newErrors.map((error) => {
      return (error.issue && error.id
        ? issueClient.updateIssue(error)
        : issueClient.createIssue(error)
      ).then((issue) => {
        return database.save({ ...error, issue })
      })
    })
  )
  return newSavedErrors
}

function distinctOccurrences(occurences: Error[]) {
  return occurences
    .sort((a, b) => a.timestamp - b.timestamp)
    .filter(
      (error, index, errors) => errors[index - 1]?.timestamp !== error.timestamp
    )
}
function zipErrors<T extends FederatedErrors>(
  errors: T[],
  compareError: (source: T, target: T) => number
): T[] {
  const changedErrors: { [key: string]: boolean } = {}
  return errors
    .sort(compareError)
    .reduce<T[]>((previous, current) => {
      const last = previous[previous.length - 1]
      if (last && compareError(last, current) === 0) {
        const previousLength = last.occurrences.length
        last.occurrences = distinctOccurrences(
          last.occurrences.concat(current.occurrences)
        )
        last.firstOccurrenceTimeStamp = Math.min(
          last.firstOccurrenceTimeStamp,
          current.firstOccurrenceTimeStamp
        )
        last.lastOccurrenceTimeStamp = Math.max(
          last.lastOccurrenceTimeStamp,
          current.lastOccurrenceTimeStamp
        )
        changedErrors[last.id || last.name] =
          previousLength != last.occurrences.length
        return previous
      } else {
        changedErrors[current.id || current.name] = !current.id
        return [...previous, current]
      }
    }, [])
    .filter((e) => changedErrors[e.id || e.name])
}
