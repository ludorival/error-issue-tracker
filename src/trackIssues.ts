import {
  EventHandler,
  Hook,
  Issue,
  TrackIssueOptions,
  TrackedIssue,
  Comparator,
} from '.'

export async function trackIssues<T extends Issue<R>, R>(
  occurrences: R[],
  {
    issueClient,
    hooks: {
      initializeNewIssue = (occurrence: R) =>
        ({ occurrences: [occurrence] } as T),
      ...rest
    },
    eventHandler = {},
  }: TrackIssueOptions<T, R>
): Promise<TrackedIssue<T>[]> {
  if (!occurrences.length) return []
  const newIssues: T[] = occurrences.map((occurrence) =>
    initializeNewIssue(occurrence)
  )

  const savedIssues: T[] = await issueClient.fetchIssues()

  const bundledIssues = bundleIssues(savedIssues, newIssues, rest, eventHandler)

  const trackedIssues = await Promise.all(
    bundledIssues.map((issue) => {
      return issue.id
        ? issueClient.updateIssue(issue as TrackedIssue<T>)
        : issueClient.createIssue(issue)
    })
  )

  return trackedIssues
}

function bundleIssues<T extends Issue<R>, R>(
  savedIssues: T[],
  newIssues: T[],
  {
    shouldBundleIssueInto = () => true,
    compareOccurrence,
    compareIssue = defaultComparator(compareOccurrence),
    getIdentifier,
  }: Hook<T, R>,
  {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onIgnoredOccurrence = () => {},
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onBundledIssue = () => {},
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onMatchedTrackedIssues = () => {},
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onNotBundledIssue = () => {},
  }: EventHandler<T, R>
): T[] {
  const mapBundled: { [key: string]: boolean } = {}
  return savedIssues
    .concat(newIssues)
    .sort(compareIssue)
    .reduce<T[]>((accumulated, current) => {
      const last = accumulated[accumulated.length - 1]
      const similar = last && compareIssue(last, current) === 0
      const shouldBundle = shouldBundleIssueInto(current, last)
      const newOccurrences =
        similar && last && shouldBundle
          ? getIdentifier
            ? keepNewOccurrences<T, R>(last, current, getIdentifier)
            : current.occurrences
          : []

      if (similar && last?.id && current.id) {
        onMatchedTrackedIssues(last, current)
      } else if ((similar && !newOccurrences.length) || !shouldBundle) {
        onIgnoredOccurrence(current, last)
        return accumulated
      } else if (similar && newOccurrences.length) {
        last.occurrences = last.occurrences.concat(newOccurrences)
        if (last.id) mapBundled[last.id] = true
        onBundledIssue(last, newOccurrences)
        return accumulated
      } else if (last) {
        onNotBundledIssue(current, last)
      }
      return [...accumulated, current]
    }, [])
    .filter((issue) => !issue.id || mapBundled[issue.id])
}

function defaultComparator<T extends Issue<R>, R>(
  compare?: Comparator<R>
): Comparator<T> {
  if (!compare)
    throw new Error('Compare Occurrence or Compare Issue must be implemented')
  return (source, target) => {
    const firstOccurrence = target.occurrences[0]
    return source.occurrences.some((o) => compare(o, firstOccurrence) === 0)
      ? 0
      : compare(source.occurrences[0], firstOccurrence)
  }
}

type ExtendedOccurrence<R> = { target: boolean; occurrence: R }
const keepNewOccurrences = <T extends Issue<R>, R>(
  fromIssue: T,
  toIssue: T,
  getIdentifier: (o: R) => number | string
): R[] => {
  const extendedTargetOccurences = toIssue.occurrences.map((occurrence) => ({
    target: true,
    occurrence,
  }))
  const extendedFromOccurences = fromIssue.occurrences.map((occurrence) => ({
    target: false,
    occurrence,
  }))

  const compareStringOrNumber = (
    a: number | string,
    b: number | string
  ): number => {
    if (typeof a === 'number' && typeof b === 'number') return a - b
    else return a.toString().localeCompare(b.toString())
  }
  const compare: Comparator<ExtendedOccurrence<R>> = (a, b) =>
    compareStringOrNumber(
      getIdentifier(a.occurrence),
      getIdentifier(b.occurrence)
    )
  const newOnes = extendedFromOccurences
    .concat(extendedTargetOccurences)
    .sort(compare)
    .reduce<ExtendedOccurrence<R>[]>((acc, value, i, array) => {
      if (array[i + 1] && compare(value, array[i + 1]) === 0) {
        return acc
      }
      if (array[i - 1] && compare(value, array[i - 1]) === 0) {
        return acc
      }
      if (!value.target) {
        return acc
      }
      return acc.concat(value)
    }, [])
    .map((e) => e.occurrence)
  return newOnes
}
