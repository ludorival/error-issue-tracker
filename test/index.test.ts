import { difference } from 'lodash'
import { Comparator, Error, trackErrors } from '../src/index'
import { anError, initOptions } from './generator'

describe('Track New Errors', () => {
  test('should track a new error', async () => {
    // given
    const errors: Error[] = [anError('A new error')]
    const options = await initOptions()
    // when
    const trackedErrors = await trackErrors(errors, options)
    // then
    expect(options.database.save).toHaveBeenCalled()
    expect(options.issueClient.createIssue).toHaveBeenCalled()
    expect(trackedErrors).toMatchSnapshot()
  })

  test('should zip sames new errors together', async () => {
    // given
    const errors: Error[] = [
      anError('A new error'),
      anError('Another error'),
      anError('A new error'),
      anError('A new error'),
      anError('Another error'),
      anError('Still an error'),
    ]
    const options = await initOptions()
    // when
    const trackedErrors = await trackErrors(errors, options)
    // then
    expect(trackedErrors).toHaveLength(3)
    expect(options.database.save).toHaveBeenCalled()
    expect(options.issueClient.createIssue).toHaveBeenCalled()
    expect(trackedErrors).toMatchSnapshot()
  })
})

describe('Track Existing Errors', () => {
  test('should track an existing error', async () => {
    // given
    const options = await initOptions([anError('A new error')])
    const errors: Error[] = [anError('A new error')]

    // when
    const trackedErrors = await trackErrors(errors, options)
    // then
    expect(trackedErrors).toHaveLength(1)
    expect(trackedErrors[0].occurrences).toHaveLength(2)
    expect(options.database.save).toHaveBeenCalled()
    expect(options.issueClient.createIssue).toHaveBeenCalledTimes(0)
    expect(options.issueClient.updateIssue).toHaveBeenCalled()
    expect(trackedErrors).toMatchSnapshot()
  })

  test('should track multiples existing errors', async () => {
    // given
    const options = await initOptions([
      anError('Error 1'),
      anError('Error 2'),
      anError('Error 1'),
      anError('Error 2'),
      anError('Error 3'),
    ])
    const errors: Error[] = [
      anError('Error 1'),
      anError('Error 2'),
      anError('Error 1'),
      anError('Error 4'),
      anError('Error 4'),
    ]

    // when
    const trackedErrors = await trackErrors(errors, options)
    // then
    expect(trackedErrors).toHaveLength(3)
    expect(options.database.save).toHaveBeenCalledTimes(3)
    expect(options.issueClient.createIssue).toHaveBeenCalledTimes(1)
    expect(options.issueClient.updateIssue).toHaveBeenCalledTimes(2)
    expect(trackedErrors).toMatchSnapshot()
  })

  test('should not track errors with same timestamp', async () => {
    // given
    const options = await initOptions([
      anError('Error 1', 1),
      anError('Error 2', 2),
    ])
    const errors: Error[] = [anError('Error 1', 1), anError('Error 2', 2)]

    // when
    const trackedErrors = await trackErrors(errors, options)
    // then
    expect(trackedErrors).toHaveLength(0)
    expect(options.database.save).toHaveBeenCalledTimes(0)
    expect(options.issueClient.createIssue).toHaveBeenCalledTimes(0)
    expect(options.issueClient.updateIssue).toHaveBeenCalledTimes(0)
  })
})

describe('Custom comparator', () => {
  test('should use a custom comparator based on differences threshold', async () => {
    // given
    const comparator: Comparator = (a, b) => {
      const threshold =
        (a.message.length -
          difference(a.message.split(''), b.message.split('')).length) /
        a.message.length
      return threshold >= 0.8 ? 0 : a.message.localeCompare(b.message)
    }
    const options = await initOptions([], comparator)

    // when
    const trackedErrors = await trackErrors(
      [anError('Error 1 : Arg0'), anError('Error 1 : Arg1')],
      options
    )

    // then
    expect(trackedErrors).toHaveLength(1)
    expect(trackedErrors).toMatchSnapshot()
  })

  test('should use a custom comparator based on an error attribute', async () => {
    // given
    const comparator: Comparator = (a, b) =>
      (a.myAttribute as number) - (b.myAttribute as number)
    const options = await initOptions([], comparator)

    // when
    const trackedErrors = await trackErrors(
      [
        { ...anError('Error 1'), myAttribute: 1 },
        { ...anError('Error 1'), myAttribute: 2 },
        { ...anError('Error 1'), myAttribute: 1 },
      ],
      options
    )

    // then
    expect(trackedErrors).toHaveLength(2)
    expect(trackedErrors).toMatchSnapshot()
  })
})
