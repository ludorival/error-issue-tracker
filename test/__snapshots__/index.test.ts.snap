// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom comparator should use a custom comparator based on an error attribute 1`] = `
Array [
  Object {
    "firstOccurrenceTimeStamp": 1,
    "id": "error-1",
    "issue": Object {
      "body": "Found 2 occurences of \\"Error 1\\"",
      "id": "1",
      "title": "Error 1",
      "url": "https://issue-tracker/project/test/1",
    },
    "lastOccurrenceTimeStamp": 3,
    "name": "Error 1",
    "occurrences": Array [
      Object {
        "message": "Error 1",
        "myAttribute": 1,
        "timestamp": 1,
      },
      Object {
        "message": "Error 1",
        "myAttribute": 1,
        "timestamp": 3,
      },
    ],
    "projectId": "test",
  },
  Object {
    "firstOccurrenceTimeStamp": 2,
    "id": "error-1",
    "issue": Object {
      "body": "Found 1 occurences of \\"Error 1\\"",
      "id": "2",
      "title": "Error 1",
      "url": "https://issue-tracker/project/test/2",
    },
    "lastOccurrenceTimeStamp": 2,
    "name": "Error 1",
    "occurrences": Array [
      Object {
        "message": "Error 1",
        "myAttribute": 2,
        "timestamp": 2,
      },
    ],
    "projectId": "test",
  },
]
`;

exports[`Custom comparator should use a custom comparator based on differences threshold 1`] = `
Array [
  Object {
    "firstOccurrenceTimeStamp": 1,
    "id": "error-1-:-arg0",
    "issue": Object {
      "body": "Found 2 occurences of \\"Error 1 : Arg0\\"",
      "id": "1",
      "title": "Error 1 : Arg0",
      "url": "https://issue-tracker/project/test/1",
    },
    "lastOccurrenceTimeStamp": 2,
    "name": "Error 1 : Arg0",
    "occurrences": Array [
      Object {
        "message": "Error 1 : Arg0",
        "timestamp": 1,
      },
      Object {
        "message": "Error 1 : Arg1",
        "timestamp": 2,
      },
    ],
    "projectId": "test",
  },
]
`;

exports[`Track Existing Errors should track an existing error 1`] = `
Array [
  Object {
    "firstOccurrenceTimeStamp": 1,
    "id": "a-new-error",
    "issue": Object {
      "body": "Found 1 occurences of \\"A new error\\"",
      "comments": Array [
        Object {
          "message": "A new error",
          "timestamp": 2,
        },
      ],
      "id": "1",
      "title": "A new error",
      "url": "https://issue-tracker/project/test/1",
    },
    "lastOccurrenceTimeStamp": 2,
    "name": "A new error",
    "occurrences": Array [
      Object {
        "message": "A new error",
        "timestamp": 1,
      },
      Object {
        "message": "A new error",
        "timestamp": 2,
      },
    ],
    "projectId": "test",
  },
]
`;

exports[`Track Existing Errors should track multiples existing errors 1`] = `
Array [
  Object {
    "firstOccurrenceTimeStamp": 1,
    "id": "error-1",
    "issue": Object {
      "body": "Found 2 occurences of \\"Error 1\\"",
      "comments": Array [
        Object {
          "message": "Error 1",
          "timestamp": 3,
        },
        Object {
          "message": "Error 1",
          "timestamp": 6,
        },
        Object {
          "message": "Error 1",
          "timestamp": 8,
        },
      ],
      "id": "1",
      "title": "Error 1",
      "url": "https://issue-tracker/project/test/1",
    },
    "lastOccurrenceTimeStamp": 8,
    "name": "Error 1",
    "occurrences": Array [
      Object {
        "message": "Error 1",
        "timestamp": 1,
      },
      Object {
        "message": "Error 1",
        "timestamp": 3,
      },
      Object {
        "message": "Error 1",
        "timestamp": 6,
      },
      Object {
        "message": "Error 1",
        "timestamp": 8,
      },
    ],
    "projectId": "test",
  },
  Object {
    "firstOccurrenceTimeStamp": 2,
    "id": "error-2",
    "issue": Object {
      "body": "Found 2 occurences of \\"Error 2\\"",
      "comments": Array [
        Object {
          "message": "Error 2",
          "timestamp": 4,
        },
        Object {
          "message": "Error 2",
          "timestamp": 7,
        },
      ],
      "id": "2",
      "title": "Error 2",
      "url": "https://issue-tracker/project/test/2",
    },
    "lastOccurrenceTimeStamp": 7,
    "name": "Error 2",
    "occurrences": Array [
      Object {
        "message": "Error 2",
        "timestamp": 2,
      },
      Object {
        "message": "Error 2",
        "timestamp": 4,
      },
      Object {
        "message": "Error 2",
        "timestamp": 7,
      },
    ],
    "projectId": "test",
  },
  Object {
    "firstOccurrenceTimeStamp": 9,
    "id": "error-4",
    "issue": Object {
      "body": "Found 2 occurences of \\"Error 4\\"",
      "id": "9",
      "title": "Error 4",
      "url": "https://issue-tracker/project/test/9",
    },
    "lastOccurrenceTimeStamp": 10,
    "name": "Error 4",
    "occurrences": Array [
      Object {
        "message": "Error 4",
        "timestamp": 9,
      },
      Object {
        "message": "Error 4",
        "timestamp": 10,
      },
    ],
    "projectId": "test",
  },
]
`;

exports[`Track New Errors should track a new error 1`] = `
Array [
  Object {
    "firstOccurrenceTimeStamp": 1,
    "id": "a-new-error",
    "issue": Object {
      "body": "Found 1 occurences of \\"A new error\\"",
      "id": "1",
      "title": "A new error",
      "url": "https://issue-tracker/project/test/1",
    },
    "lastOccurrenceTimeStamp": 1,
    "name": "A new error",
    "occurrences": Array [
      Object {
        "message": "A new error",
        "timestamp": 1,
      },
    ],
    "projectId": "test",
  },
]
`;

exports[`Track New Errors should zip sames new errors together 1`] = `
Array [
  Object {
    "firstOccurrenceTimeStamp": 1,
    "id": "a-new-error",
    "issue": Object {
      "body": "Found 3 occurences of \\"A new error\\"",
      "id": "1",
      "title": "A new error",
      "url": "https://issue-tracker/project/test/1",
    },
    "lastOccurrenceTimeStamp": 4,
    "name": "A new error",
    "occurrences": Array [
      Object {
        "message": "A new error",
        "timestamp": 1,
      },
      Object {
        "message": "A new error",
        "timestamp": 3,
      },
      Object {
        "message": "A new error",
        "timestamp": 4,
      },
    ],
    "projectId": "test",
  },
  Object {
    "firstOccurrenceTimeStamp": 2,
    "id": "another-error",
    "issue": Object {
      "body": "Found 2 occurences of \\"Another error\\"",
      "id": "2",
      "title": "Another error",
      "url": "https://issue-tracker/project/test/2",
    },
    "lastOccurrenceTimeStamp": 5,
    "name": "Another error",
    "occurrences": Array [
      Object {
        "message": "Another error",
        "timestamp": 2,
      },
      Object {
        "message": "Another error",
        "timestamp": 5,
      },
    ],
    "projectId": "test",
  },
  Object {
    "firstOccurrenceTimeStamp": 6,
    "id": "still-an-error",
    "issue": Object {
      "body": "Found 1 occurences of \\"Still an error\\"",
      "id": "6",
      "title": "Still an error",
      "url": "https://issue-tracker/project/test/6",
    },
    "lastOccurrenceTimeStamp": 6,
    "name": "Still an error",
    "occurrences": Array [
      Object {
        "message": "Still an error",
        "timestamp": 6,
      },
    ],
    "projectId": "test",
  },
]
`;
